torch/nn/parallel/distributed_c10d.py

```diff
@@ -0,0 +1,375 @@
+                for tensor, param_data in zip(tensors, module_params_data):
+                    param_data.set_(tensor)
+
+        # module buffer sync
+        if self.broadcast_buffers:
+            if len(self.modules_buffers_data[0]) > 0:
+                # cross-node buffer sync
+                self._dist_broadcast_coalesced(self.modules_buffers_data[0],
+                                               self.broadcast_bucket_size)
+                if len(self.device_ids) > 1:
+                    # intra-node buffer sync
+                    result = broadcast_coalesced(self.modules_buffers_data[0],
+                                                 self.device_ids,
+                                                 self.broadcast_bucket_size)
+                    for tensors, module_buffers_data in zip(result[1:], self.modules_buffers_data[1:]):
+                        for tensor, buffer_data in zip(tensors, module_buffers_data):
+                            buffer_data.set_(tensor)
+
+    def _register_grad_hooks(self):
+        self._grad_accs = []  # need to keep them in scope
+        for device_idx, module in enumerate(self._module_copies):
+            for p in module.parameters():
+                if p.requires_grad:
+                    p_tmp = p.expand_as(p)
+                    grad_acc = p_tmp.grad_fn.next_functions[0][0]
+                    grad_acc.register_hook(self._make_param_hook(p, device_idx))
+                    self._grad_accs.append(grad_acc)
+
+    def _make_param_hook(self, param, device_idx):
+        bucket_idx, bucket_offset = self.bucket_map[param]
+
+        def distributed_data_parallel_hook(*unused):
+            if param.grad.requires_grad:
+                raise RuntimeError("DistributedDataParallelC10d only works "
+                                   "with gradients that don't require grad")
+            bucket = self.buckets[bucket_idx][device_idx]
+            bucket[bucket_offset] = param.grad.data
+            self.buckets_ready_size[bucket_idx][device_idx] += 1
+
+            # We can flush these and save memory for replicas
+            if device_idx > 0:
+                param.grad = None
+                param.data.set_()
+
+            # Current device's bucket is full
+            if self.buckets_ready_size[bucket_idx][device_idx] == self.bucket_sizes[bucket_idx]:
+                self.devs_ready[bucket_idx] += 1
+                if self.devs_ready[bucket_idx] < len(self.device_ids):
+                    return
+
+                # Now all devices's buckets with index: bucket_idx are ready
+                if bucket_idx == self.next_bucket:
+                    self._queue_reduction(bucket_idx)
+                    self.next_bucket -= 1
+                    # Now reduce anything that is ready but not yet reduced
+                    if len(self.ready_buckets_not_reduced) > 0:
+                        sorted_todo = sorted(self.ready_buckets_not_reduced, reverse=True)
```

Comment:
FWIW this introduces a O(n^2 log n) complexity where n is the number of buckets. It would be much simpler and faster to keep an array of True/False flags for each bucket that tells you if it's ready or not. Then, once you get here, you start at `self.next_bucket` and iterate over the array until you reach a False, or it ends.